File: Leetcode/JS/Leetcode859-亲密字符串.js
var buddyStrings = function(s, goal) {
    if(s.length != goal.length) return false;
    if(s === goal){
        let mp = new Map();
        for(const ch of s){
            mp.set(ch,(mp.get(ch) || 0) + 1);
            if(mp.get(ch) > 1) return true;
        }
        return false;
    }
    let arr = new Array();
    for(let i = 0;i < s.length;i++){
        if(s[i] != goal[i]) arr.push(i);
    }
    if(arr.length != 2) return false;
    let [x,y] = [arr[0],arr[1]];
    return s[x] == goal[y] && s[y]== goal[x];
};

File: Leetcode/PY/Leetcode859-亲密字符串.py
class Solution:
    def buddyStrings(self, s: str, goal: str) -> bool:
        if len(s) != len(goal): return False
        if s == goal:
            mp = defaultdict(int)
            for ch in s:
                mp[ch]+=1
                if mp[ch] > 1: return True
            return False
        idx = []
        for i in range(len(s)):
            if s[i] != goal[i]:
                idx.append(i)
        if len(idx) != 2: return False
        x,y = idx[0],idx[1]
        return s[x] == goal[y] and s[y] == goal[x]
